src/System.Net.Security/src/System/Net/SecureProtocols/_SslState.cs

```diff
@@ -0,0 +1,1745 @@
+            {
+                return;
+            }
+
+            lock (this)
+            {
+                LazyAsyncResult ar = _QueuedReadStateRequest as LazyAsyncResult;
+                if (ar != null)
+                {
+                    _QueuedReadStateRequest = null;
+                    ar.InvokeCallback(renegotiateBuffer);
+                }
+                else
+                {
+                    AsyncProtocolRequest request = (AsyncProtocolRequest)_QueuedReadStateRequest;
+                    request.Buffer = renegotiateBuffer;
+                    _QueuedReadStateRequest = null;
+                    ThreadPool.QueueUserWorkItem(new WaitCallback(AsyncResumeHandshakeRead), request);
+                }
+            }
+        }
+
+        // Returns: 
+        // true  - operation queued
+        // false - operation can proceed
+        internal bool CheckEnqueueWrite(AsyncProtocolRequest asyncRequest)
+        {
+            // Clear previous request.
+            _QueuedWriteStateRequest = null;
+            int lockState = Interlocked.CompareExchange(ref _LockWriteState, LockWrite, LockNone);
+            if (lockState != LockHandshake)
+            {
+                // Proceed with write.
+                return false;
+            }
+            LazyAsyncResult lazyResult = null;
+            lock (this)
+            {
+                if (_LockWriteState != LockHandshake)
+                {
+                    // Handshake has completed before we grabbed the lock.
+                    CheckThrow(true);
+                    return false;
+                }
+
+                _LockWriteState = LockPendingWrite;
+
+                // Still pending, wait or enqueue.
+                if (asyncRequest != null)
+                {
+                    _QueuedWriteStateRequest = asyncRequest;
+                    return true;
+                }
+                lazyResult = new LazyAsyncResult(null, null, /*must be */null);
+                _QueuedWriteStateRequest = lazyResult;
+            }
+
+            // Need to exit from lock before waiting.
+            lazyResult.InternalWaitForCompletion();
+            CheckThrow(true);
+            return false;
+        }
+
+        internal void FinishWrite()
+        {
+            int lockState = Interlocked.CompareExchange(ref _LockWriteState, LockNone, LockWrite);
+            if (lockState != LockHandshake)
+            {
+                return;
+            }
+            lock (this)
+            {
+                object obj = _QueuedWriteStateRequest;
+                if (obj == null)
+                {
+                    // A repeated call.
+                    return;
+                }
+
+                _QueuedWriteStateRequest = null;
+                if (obj is LazyAsyncResult)
+                {
+                    // Sync handshake is waiting on other thread.
+                    ((LazyAsyncResult)obj).InvokeCallback();
```

Comment:
It's ok that the callback is being invoked while holding the lock?  No opportunities for deadlock as a result?